Fixes vent crawl, fixes xenomorph HUD, adds ability for hunters to pry open doors.

Kearel opened the pull request:
The problem with vent crawl was that when it was checking the contents of the mob against the different inventory slots (which is a good way of doing it) it would trust that isnull would return 1 if no slot was found. But isnull of 0 is still 0. Meaning body parts, organs, embedded objects, swallowed people, they'd all cause ventcrawling to fail.

Xenomorph HUD was bugged. Didn't have icons set in the right places as it still used the old way of how HUDs were set up.

Door prying was added in an attempt to emulate old Xenomorph hunter behavior since hunters can't acid spit doors and moving from vent to vent is a really bad experience due to never really knowing what vent in a room you'd end up at.

:cl: Kearel
add: Xenomorph hunters can now pry open airlocks that are not bolted/welded shut.
/:cl:

PsiOmegaDelta commented on code/modules/mob/living/carbon/human/species/xenomorphs/alien_powers.dm:
```diff
@@ -261,4 +261,30 @@ mob/living/carbon/human/proc/xeno_infest(mob/living/carbon/human/M as mob in ovi
+
+	src.visible_message("\The [src] begins to pry open \the [A]!")
+
+	if(!do_after(src,80,A))
```
8 seconds to breach an airlock, bolted or not, seems a bit short but other than that I suppose it's a good ability.

atlantiscze replied:
I'd be for keeping it at 8 seconds, but add a requirement for the airlock to be broken first.

Raptor1628 commented:
Nice

Kearel commented:
This need anything else? Any issues?

PsiOmegaDelta commented on code/modules/mob/living/carbon/human/species/xenomorphs/alien_powers.dm:
```diff
@@ -261,4 +261,30 @@ mob/living/carbon/human/proc/xeno_infest(mob/living/carbon/human/M as mob in ovi
+	src.visible_message("\The [src] begins to pry open \the [A]!")
+
+	if(!do_after(src,120,A))
+		return
```
Should re-check density with the do_after() as the door may have opened in the meantime.

PsiOmegaDelta commented:
No other issues noted. Double-checked and confirmed that even if one force-open doors, bolts still prevent you from doing so.

Kearel commented:
The can_open proc will always check if the door is welded/bolted regardless if the proc is forced or not, so open() won't do anything there. I'll put in a check for that so the door opening message doesn't happen if its still closed.

Kearel commented:
the abiilty rework commit was out of the scope for this PR so I reverted it. I'll come back to it later.

I did some testing to make sure and yes, door bolts will make door pry fail. Reworked the message a bit and checked for density like you asked so it won't close on them.

Also splitted the ALIENS_ALLOWED config option, it seemed kind of weird to me that it dictated both the player's ability to lay eggs and the ability for random xeno spawns. It defaults to off so it won't affect anything unless the server config gets changed (or if somebody presses the toggle)

PsiOmegaDelta commented on code/modules/admin/admin.dm:
```diff
@@ -817,6 +817,16 @@ proc/admin_notice(var/message, var/rights)
+/datum/admins/proc/toggle_alien_eggs()
+	set category = "Server"
+	set desc="Toggle xenomorph egg laying"
+	set name="Toggle Alien Eggs"
```
Should include the rights check
```
if(!check_rights(R_ADMIN))
    return
```

Kearel commented:
Should I add rights checks to the other alien toggle proc? I noticed it didn't have one.

Raptor1628 commented:
Yes please.

Kearel commented:
Alright there.

PsiOmegaDelta commented:
:+1:

mwerezak commented:
:+1: for fixing the issue. Just want to make a note though.

So when something ventcrawls, the game loops over every item in it's `contents` and then calls `is_allowed_vent_crawl_item()` on it. That proc then calls `get_inventory_slot()`, which loops over every item equipped by the mob *again* to check `if(get_equipped_item(slot) == I)`. Which does so by jumping into a giant switch statement to get the item in that slot.

That is hilariously inefficient, hopefully we'll remember to deal with that sometime.

Kearel commented:
Might be more efficient to check stomach contents/shrapnel contents/organ contents since those are what I think are most of the problem.

mwerezak commented:
Hm. Maybe we could just check if `get_equipped_items()` returns anything at all.

Kearel commented:
Yeah, a singular check through all the slots is actually a lot better since its a finite list and a small(ish) one that that.
